src/alchemlyb/tests/test_preprocessing.py

```diff
@@ -68,6 +68,29 @@ def slicer(self, *args, **kwargs):
     def test_basic_slicing(self, data, size):
         assert len(self.slicer(data, lower=1000, upper=34000, step=5)) == size
 
+    @pytest.mark.parametrize(('data', 'lower', 'upper'),
+                             [
+                                 (gmx_benzene_dHdl(), 1000, 34000),
+                                 (gmx_benzene_u_nk(), 1000, 34000),
```

Comment:
It's not great to call the data provider functions here because this adds to the time during which tests are gathered in serial as opposed during test execution (which can be parallelized). I know that we already have plenty examples of doing exactly that elsewhere so it's really not your fault.

I am just wondering if there are better ways to do this, e.g., passing the function itself as `data_generator` and then calling it as `data = data_generator()` might be easiest. It would be better if we could generate them as fixtures (but parametrized fixtures are still awkward IIRC).

Reply:
@orbeckst The parametrized fixtures could be made easy with `pytest-lazy-fixture`

I have posted an example here as is noted in their website https://pypi.org/project/pytest-lazy-fixture/. This is conda installable as well.
```
@pytest.fixture(params=[
    pytest.lazy_fixture('one'),
    pytest.lazy_fixture('two')
])
def some(request):
    return request.param

@pytest.fixture
def one():
    return 1

@pytest.fixture
def two():
    return 2
```

Reply:
This makes a lot of sense @orbeckst, I wasn't aware of this issue. Unfortunately I haven't had time to look into this yet. Current version therefore still has the greedy evaluation.

Reply:
`lazy_fixture` looks like a convenient solution. I looked at the package and I am not sure how well maintained it is. There's a commit from Jan 2022 in https://github.com/tvorog/pytest-lazy-fixture but the pypi page https://pypi.org/project/pytest-lazy-fixture/ states that the highest supported Python version is 3.7. @xiki-tempula , do you have experience with it in production code? It would be annoying to rewrite quite a few of our tests, just to discover that it breaks with a new release of pytest...

Reply:
I had a look around in our tests and we do have examples of evaluating the loading functions later: see e.g. https://github.com/alchemistry/alchemlyb/blob/72622c99c16c74efd1f8517cb15ef21e16d6ebda/src/alchemlyb/tests/test_fep_estimators.py#L145-L168

This is not too bad.

Reply:
I raised #206 for fixing the other instances of upfront data loading.

Reply:
@orbeckst yes, I used the lazy_fixture in _our_ production code but the reason that I use it is that it was there. It works fine for me but I do share your concern about it not being updated for years.

On the other hand, one could also use the official pytest solution

```
@pytest.fixture(params=['dir1_fixture', 'dir2_fixture'])
def dirname(request):
    return request.getfixturevalue(request.param)
```

Reply:
This looks straightforward
```python
@pytest.mark.parametrize(('dataloader', 'size'), [('gmx_benzene_dHdl', 661), 
                                                  ('gmx_benzene_u_nk', 661)]) 
     def test_basic_slicing(self, dataloader, size, request):
         data = request.getfixturevalue(dataloader)
         assert len(self.slicer(data, lower=1000, upper=34000, step=5)) == size 
```